Use index.rst in package root as alternative to README

luator opened the pull request:
## Description

Add "index.rst" as a candidate name for the README file (rename the corresponding functions to "..._index_or_readme" to reflect this).

Rewrite the code search for the file such that in case of multiple matches, the order in which they are specified in `options` is adhered to (previously it was just randomly picking whichever was first returned by `glob("*")`).

It resolves the issue that the content in the README does not always make sense to be used as documentation main page by adding the option to provide an additional index.rst next to the README (GitHub will still show the README but the documentation will use the index.rst).

I would actually prefer to locate the index.rst in the doc/ folder as this seems cleaner. However, this would cause trouble with relative links as in the build dir, the index.rst is outside the doc/ folder.  Thus I decided to put it at the package root, so the relative path from index.rst to other doc files is the same.  This has the nice side effect that it is also much easier to implement, as I basically just needed to extend the README search a bit :)


## How I Tested

Building documentation for a package with and without an `index.rst` next to the README.



## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

luator committed:
Use index.rst in package root as alternative to README

Add "index.rst" as a candidate name for the README file (rename the
corresponding functions to "..._index_or_readme" to reflect this).

Rewrite the code search for the file such that in case of multiple
matches, the order in which they are specified in `options` is adhered
to (previously it was just randomly picking whichever was first returned
by `glob("*")`).

It resolves the issue that the content in the README does not always
make sense to be used as documentation main page by adding the option to
provide an alternative index.rst next to the README (GitHub will still
show the README but the documentation will use the index.rst).

luator committed:
Change header line style for index.rst.in

Make the main header use
```
*************************
Asterisks above and below
*************************
```
and change the Indices and License headers to use `==` instead of `--`.
This should make it more consistent with READMEs typically using `==`
for the top level header.

MaximilienNaveau commented on CHANGELOG.md:
```diff
@@ -5,6 +5,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
+### Added
+- If the package root contains a file `index.rst`, it is used instead of the README to
```
index.rst is not a super nice name... as it usually refers to the "main page" of the documentation.

Hence I would check for some doc/doc_readme.rst? or doc_readme.rst?

WDYT?

luator replied:
Well, it _is_ supposed to be the documentation main page (breathing-cat basically takes this file and just appends the stuff for API docs, etc.) and explicitly not a README (which IMO serves a different purpose).

We could think about different names, if you think `index.rst` is confusing, but I wouldn't use "readme". 
 Maybe something like `doc_mainpage.rst`?  Unfortunately it needs to be in the package root, not in `doc/` as otherwise it gets very confusing with relative links.

MaximilienNaveau replied:
ok for doc_mainpage.rst

for the doc/ folder I am not sure about it but let's keep it in the root folder for now

luator replied:
Done.
I also added support for `doc_mainpage.md` (as I realised, this is just one line to add) and refactored a bit.

luator committed:
Use "doc_mainpage.rst" instead of "index.rst"

Also rename the "index_or_readme" functions to "mainpage" to be more
generic.

luator committed:
Add support for doc_mainpage.md

luator committed:
refactor: Return target name from _copy_mainpage()

Better have the code deciding on the name only in one place.  Also
change from "readme" to "mainpage".
